### PR TITLE
Enable nullable

### PR DESCRIPTION
**Pull request by iamcarbon**
This PR

- Enables nullable
- Replaces == null with is null
- Seals a few internal classes
- Remove a single allocation by creating a static byte array

Nullable is disabled in various classes with initialization patterns that don't work well with nullable until C# v9.

### CODE REVIEW

**Comment by iamcarbon** [comment]
@adamhathcock Ready for review. 

If things look good, I'll do another PR that focused on eliminating allocations. I spotted at least 20 spots that could be optimized but didn't want this PR to get to big.

File: src/SharpCompress/Archives/AbstractArchive.cs [hunk 1]
```diff
@@ -48,12 +48,14 @@ internal AbstractArchive(ArchiveType type, IEnumerable<Stream> streams, ReaderOp
             lazyEntries = new LazyReadOnlyCollection<TEntry>(LoadEntries(Volumes));
         }
 
+#nullable disable
```
Comment (adamhathcock) [inline src/SharpCompress/Archives/AbstractArchive.cs:51]: Not too familiar but why is this necessary?

File: src/SharpCompress/Archives/GZip/GZipWritableArchiveEntry.cs [hunk 2]
```diff
@@ -1,12 +1,14 @@
-﻿using System;
+﻿#nullable disable
```
Comment (adamhathcock) [inline src/SharpCompress/Archives/GZip/GZipWritableArchiveEntry.cs:1]: Disabled for file?

File: src/SharpCompress/Archives/Rar/RarArchiveVolumeFactory.cs [hunk 3]
```diff
@@ -34,20 +34,20 @@ internal static IEnumerable<RarVolume> GetParts(FileInfo fileInfo, ReaderOptions
             {
                 yield break; //if file isn't volume then there is no reason to look
             }
-            fileInfo = GetNextFileInfo(ah, part.FileParts.FirstOrDefault() as FileInfoRarFilePart);
+            fileInfo = GetNextFileInfo(ah, part.FileParts.FirstOrDefault() as FileInfoRarFilePart)!;
```
Comment (adamhathcock) [inline src/SharpCompress/Archives/Rar/RarArchiveVolumeFactory.cs:37]: Should fileInfo just be nullable?

File: src/SharpCompress/Archives/Zip/ZipArchive.cs [hunk 4]
```diff
@@ -15,7 +15,9 @@ namespace SharpCompress.Archives.Zip
 {
     public class ZipArchive : AbstractWritableArchive<ZipArchiveEntry, ZipVolume>
     {
+#nullable disable
```
Comment (adamhathcock) [inline src/SharpCompress/Archives/Zip/ZipArchive.cs:18]: Why disable for this?  It should never be null

**Comment by adamhathcock** [comment]
I should have read your general disabling comment.  I can't wait for C#9 improving general usage.  I didn't bother turning on nullables in this because it was a lot of work.
